torch/csrc/api/src/nn/modules/batchnorm.cpp

```diff
@@ -72,5 +78,75 @@ Tensor BatchNormImpl::pure_forward(
       torch::cuda::cudnn_is_available());
 }
 
+template <size_t D, typename Derived>
+BatchNormImplBase<D, Derived>::BatchNormImplBase(const BatchNormOptionsv2<D>& options_)
+    : options(options_) {
+  if (options.affine()) {
+    weight = this->register_parameter("weight", torch::ones({options.num_features()}));
+    bias = this->register_parameter("bias", torch::zeros({options.num_features()}));
+  }
+  if (options.track_running_stats()) {
+    running_mean = this->register_buffer("running_mean", torch::zeros({options.num_features()}));
+    running_var = this->register_buffer("running_var", torch::ones({options.num_features()}));
+    num_batches_tracked = this->register_buffer("num_batches_tracked", torch::tensor(0, torch::dtype(torch::kLong)));
+  }
+  reset();
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::reset_parameters() {
+  reset_running_stats();
+  if (options.affine()) {
+    torch::nn::init::ones_(weight);
+    torch::nn::init::zeros_(bias);
+  }
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::reset_running_stats() {
+  if (options.track_running_stats()) {
+    torch::nn::init::zeros_(running_mean);
+    torch::nn::init::ones_(running_var);
+    torch::nn::init::zeros_(num_batches_tracked);
+  }
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::reset() {
+  reset_parameters();
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::BatchNorm" << D << "d("
+         << "num_features=" << options.num_features() << ", "
+         << "eps=" << options.eps() << ", "
+         << "momentum=" << options.momentum() << ", "
+         << "affine=" << options.affine() << ", "
+         << "track_running_stats=" << options.track_running_stats() << ")";
+}
+
+Tensor BatchNorm1dImpl::forward(const Tensor& input) {
+  TORCH_CHECK(
+      input.dim() != 2 && input.dim() !=3,
+      "expected 2D or 3D input (got %dD input)", input.dim());         
```

Comment:
I feel that we can add a virtual function `_check_input_dim` to the `BatchNormImplBase` class, override its implementation in `BatchNorm1dImpl`, and call this `_check_input_dim` function from `forward`, to match the Python version even better :D

Reply:
OK. I added pure virtual function `_check_input_dim` to `BatchNormImplBase` class.